client/src/main/java/io/pravega/client/stream/impl/ReaderGroupState.java

```diff
@@ -1194,12 +1261,182 @@ private void write00(ClearCheckpointsBefore object, RevisionDataOutput out) thro
+                };
+                revisionDataOutput.writeBuffer(new ByteArraySegment(object.newConfig.toBytes()));
+                revisionDataOutput.writeMap(object.startingSegments, segmentWithRangeSerializer, RevisionDataOutput::writeLong);
+                revisionDataOutput.writeMap(object.endSegments, segmentSerializer, RevisionDataOutput::writeLong);
+            }
+        }
+    }
+
+    @Builder
+    @Data
+    @EqualsAndHashCode(callSuper = false)
+    static class ReaderGroupStateResetStart extends ReaderGroupStateUpdate {
+        private final ReaderGroupConfig newConfig;
+        private final long newGeneration;
+
+        /**
+         * @see ReaderGroupState.ReaderGroupStateUpdate#update(ReaderGroupState)
+         */
+        @Override
+        void update(ReaderGroupState state) {
+        }
+
+        @Override
+        public ReaderGroupState applyTo(ReaderGroupState oldState, Revision newRevision) {
+            return new ReaderGroupState(oldState.getScopedStreamName(), oldState.getConfig(), newRevision, oldState.checkpointState,
+                    oldState.distanceToTail, oldState.futureSegments, oldState.assignedSegments, oldState.unassignedSegments,
+                    oldState.lastReadPosition, oldState.endSegments, newGeneration, ConfigState.REINITIALIZING, newConfig);
+        }
+
+        private static class ReaderGroupStateResetStartBuilder implements ObjectBuilder<ReaderGroupStateResetStart> {
+        }
+
+        private static class ReaderGroupStateResetStartSerializer
+                extends VersionedSerializer.WithBuilder<ReaderGroupStateResetStart, ReaderGroupStateResetStartBuilder> {
+            @Override
+            protected ReaderGroupStateResetStartBuilder newBuilder() {
+                return builder();
+            }
+
+            @Override
+            protected byte getWriteVersion() {
+                return 0;
+            }
+
+            @Override
+            protected void declareVersions() {
+                version(0).revision(0, this::write00, this::read00);
+            }
+
+            private void read00(RevisionDataInput in, ReaderGroupStateResetStartBuilder builder) throws IOException {
+                builder.newConfig(ReaderGroupConfig.fromBytes(ByteBuffer.wrap(in.readArray())));
+                builder.newGeneration(in.readLong());
+            }
+
+            private void write00(ReaderGroupStateResetStart object, RevisionDataOutput out) throws IOException {
+                out.writeBuffer(new ByteArraySegment(object.newConfig.toBytes()));
+                out.writeLong(object.newGeneration);
+            }
+        }
+    }
+
+    @Builder
+    @Data
+    @EqualsAndHashCode(callSuper = false)
+    public static class ChangeConfigState extends ReaderGroupStateUpdate {
+        private final ConfigState newConfigState;
+        private final long newGeneration;
+
+        @Override
+        void update(ReaderGroupState state) {
```

Comment:
Why did we choose to override applyTo() instead of update()?

Reply:
`applyTo()` allowed for returning the entire state. `applyTo()` uses `update()` by default so it doesn't really make a difference. `applyTo()` allowed for changing the generation and configState without having setters. I did not want to have a setter for the generation, so that the user cannot change it unless he makes a configState update also.